### ISSUE TITLE
InvalidArgumentError : input must be 4-dimensional[100,100,3]

### ISSUE THREAD

**Issue by civilmanxx** [tensorflow/tensorflow#5125]
I am trying to use inception V1 model to predict single image. The example on the tensor flow website uses inception V3, which needs 1.7s-1.9s on a single Tian X GPU. This is not fast. 

I use the following python code to test V3 performance, which by changing **classify_image.py**

     #... above code is for load graph and read image

    t1=time.time()

    softmax_tensor = sess.graph.get_tensor_by_name('softmax:0') 
    predictions = sess.run(softmax_tensor,{'DecodeJpeg/contents:0': image_data}) 
    predictions= np.squeeze(fpredictions)

    delta=time.time()-t1
    print(delta)

I think the major reason is V3 has 42 layers, which is too large. 

However, when i change the above code for V1 model. I meet errors. First I get the node names from V1 model .pb file as below:
**input
conv2d0_w
conv2d0_b
conv2d1_w**
...
...
...
**softmax2
output
output1
output2**

So the entry point of V1 model is **input**. The entry point of V3 is **DecodeJpeg/contents**, which accept image buffer from **tf.gfile.FastGFile(image, 'rb').read()** function directly. But V1 does not. So i use the code below to get the image and return as an array.

**image_data**=**cv2**.imread(image)  

Now the code for V1 model become for prediction become

    softmax_tensor = sess.graph.get_tensor_by_name('softmax2:0')
    predictions = sess.run(softmax_tensor, {'input:0': image_data})

However, I get error below when i run the code.

InvalidArgumentError (see above for traceback): **input must be 4-dimensional**[100,100,3]
	 [[Node: conv2d0_pre_relu/conv = Conv2D[T=DT_FLOAT, data_format="NHWC", padding="SAME", strides=[1, 2, 2, 1], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_input_0, conv2d0_w)]]

It says it need a 4-dimensional input. My image array has width, height and channel only. So how can i create a 4D tensor from a single image for V1 model as input data?

**Comment by civilmanxx** [comment]
As the data format is **NHWC**, It seems that the first dim "N" means batch size. So if the single input image is 100*100*3, then how to add 1 for "N", add a 100*100*1=1 on top of image?

**Comment by asimshankar** [comment]
@civilmanxx : A couple of things:

- IIUC, the inception v1 model requires the input to be normalized. Which means that the images must be of a precise size.
- You're right, the 4 dimensions is because the model accepts batches of images
- You can see this normalization in the [label_image C++ example](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image) ([main.cc](https://github.com/tensorflow/tensorflow/blob/20c02ffe0e6b8af4902487f852e15daa16caa523/tensorflow/examples/label_image/main.cc#L117)) - which casts the pixels to a float, resizes it, normalized the value by subtracting a mean and extends the number of dimensions.

Hope that helps. Please feel free to reopen the bug if your concerns weren't addressed.

**Comment by RitiP** [comment]
@civilman628 Did you find a solution for this? I'm facing the same issue.
